### PR TITLE
Mongo Migration: Add handler to transfer all recipes to mongo from data.go

### PR DESCRIPTION
**Pull request by justinpjose**
### What

**Describe what you have changed and why.**
1.  Add "AddRecipe" to mongo package to add recipe
2. Add "AddAllRecipeHandler" to use "AddRecipe" function to transfer all recipes to mongo
3. Mock to test the functionality

### How to review

**Describe the steps required to test the changes.**
Run the tests in the handlers_test.go for "AddAllRecipes"

### Who can review

**Describe who worked on the changes, so that other people can review.**
Justin worked on the changes

### CODE REVIEW

File: go.sum [hunk 1]
```diff
@@ -1,23 +1,43 @@
 github.com/BurntSushi/toml v0.3.1/go.mod h1:xHWCNGjB5oqiDr8zfno3MHue2Ht5sIBksp03qcyfWMU=
 github.com/ONSdigital/dp-api-clients-go v0.0.0-20190920133223-0b75bbb235dd h1:GKyefMVR8NpBNYkdo4AS7K6zRDIw1MQ/hEVRH3s1CEc=
 github.com/ONSdigital/dp-api-clients-go v0.0.0-20190920133223-0b75bbb235dd/go.mod h1:vx3olkUJib+ZtLL6njP8iX1fjePokXyIcitFWKQ9F/A=
+github.com/ONSdigital/dp-authorisation v0.0.0-20191018110224-d4543527d6cb h1:k58Fs56RCIHAL3GToXYwO3IfKKyIN2JvvkQHANpIZ2Q=
```
Comment (DavidSubiros) [inline go.sum:4]: can you run `go mod tidy` please?

File: api/api.go [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+		Router:            router,
+		EnableMongoData:   cfg.MongoConfig.EnableMongoData,
+		EnableMongoImport: cfg.MongoConfig.EnableMongoImport,
+	}
+
+	api.get("/health", api.HealthCheck)
+	api.get("/recipes", api.RecipeListHandler)
+	api.get("/recipes/{id}", api.RecipeHandler)
+	if api.EnableMongoImport {
+		api.Router.HandleFunc("/allrecipes", api.AddAllRecipeHandler).Methods("POST")
```
Comment (DavidSubiros) [inline api/api.go:51]: I don't know if this belongs to another PR or this but I would suggest to keep the `/recipes` path here (instead of allrecipes) -

Comment (DavidSubiros) [inline api/api.go:51]: I would suggest using the endpoint `/recipes` here (we can have different functions for different methods for the same endpoint)

File: mongo/mongo.go [hunk 3]
```diff
@@ -27,3 +35,50 @@ func (m *Mongo) Init() (session *mgo.Session, err error) {
...
+			return nil, errs.ErrRecipeNotFound
+		}
+		return nil, err
+	}
+
+	return &recipe, nil
+}
+
+//AddRecipe adds a recipe document
+func (m *Mongo) AddRecipe(item recipe.Response) error {
```
Comment (DavidSubiros) [inline mongo/mongo.go:79]: Is there any way to add multiple recipes in a single transaction in MongoDB? If it's easy I would suggest doing that to improve performance.

I had a look at the code in GlobalSign library and I don't think it's that trivial, so maybe this should be part of another task.

File: api/handlers.go [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+}
+
+//HealthCheck - health check endpoint
+func (api *RecipeAPI) HealthCheck(w http.ResponseWriter, req *http.Request) {
+	// Set status to 200 OK
+	w.WriteHeader(200)
+}
+
+//AddAllRecipeHandler - Adds all the recipes from data.go to the mongo database
+//USAGE: curl -X POST http://localhost:22300/allrecipes
```
Comment (DavidSubiros) [inline api/handlers.go:99]: if we decide to use `/recipes` endpoint, we should also update this comment.

File: api/handlers.go [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+func (api *RecipeAPI) AddAllRecipeHandler(w http.ResponseWriter, req *http.Request) {
+	for _, item := range recipe.FullList.Items {
+		err := api.dataStore.Backend.AddRecipe(item)
+		if err != nil {
+			log.Event(req.Context(), "error in adding all recipes to mongo from data.go", log.ERROR, log.Error(err))
+			w.WriteHeader(http.StatusInternalServerError)
+			return
```
Comment (DavidSubiros) [inline api/handlers.go:106]: maybe instead of returning here, we can complete the loop, keep track of any error, and if there is any error, write the InternalServerError at the end?
Something like:
```
hasErrors := false
for _, item := range recipe.FullList.Items {
    ...
    if err != nil {
        hasErrors = true
    }
...

if HasErrors {
    w.WriteHeader(http.StatusInternalServerError)
}
```


Another option (with better performance) would be to have a single query against mongodb to add all the items in a single query, if it is possible at all.

File: cmd/dp-recipe-api/main.go [hunk 6]
```diff
@@ -37,86 +43,58 @@ func main() {
...
-	w.Write(b)
-}
+	apiErrors := make(chan error, 1)
 
-func recipeHandler(w http.ResponseWriter, req *http.Request) {
-	vars := mux.Vars(req)
-	recipeID := vars["id"]
-	logD := log.Data{"recipe_id": recipeID}
+	// Gracefully shutdown the application closing any open resources.
+	gracefulShutdown := func() {
```
Comment (DavidSubiros) [inline cmd/dp-recipe-api/main.go:72]: I don't know if graceful shutdown is part of this PR, feel free to ignore this comment otherwise.

If we want to check the timeout we can create a separate go-routine here, which uses the shutdown context, and calls cancel() at the end (with a defer), then the main thread would wait for shutdownContext.Done(), and exit if there is no error.
An example would be the code in `dp-dataset-api`

Comment (justinpjose) [reply cmd/dp-recipe-api/main.go:72]: Different PR

File: api/handlers.go (hunk 5 above)
Comment (justinpjose) [reply api/handlers.go:106]: But wouldn't that be inefficient? If an error is found in one of the recipe items, then it would continue to iterate over the remaining items the recipe list even if hasErrors has been set to true 

On the other hand, the return statement within the for loop would exit the loop when an error is found. 

Am I right in this?

File: go.sum (hunk 1 above)
Comment (justinpjose) [reply go.sum:4]: go.mod and go.sum has been tidied in the latest commit

File: api/api.go (hunk 2 above)
Comment (justinpjose) [reply api/api.go:51]: The `/allrecipes` is an one-off endpoint. This endpoint is used to transfer all the existing recipes into mongo from the data.go file (where the recipes are all hard-coded). The migration to mongo has taken place, this endpoint will no longer be useful. 

Also, I am using `/recipes` endpoint with POST method to add recipe in mongo in another PR which i haven't send in `dev_code_review` as it is blocked because I need to do audit and authentication as it is dealing with POST request. Once the audit library has been fully updated, then i will complete the PR and push it for review

Comment (justinpjose) [reply api/api.go:51]: Reasoning mentioned in previous comment

File: mongo/mongo.go (hunk 3 above)
Comment (justinpjose) [reply mongo/mongo.go:79]: Adding multiple recipes in a single transaction would be mentioned in an upcoming task when an UI would be created to do

File: api/handlers.go (hunk 4 above)
Comment (justinpjose) [reply api/handlers.go:99]: Reasoning mentioned in previous comment

File: api/api.go (hunk 2 above)
Comment (DavidSubiros) [reply api/api.go:51]: ok thanks for the explanation!

File: api/handlers.go (hunk 5 above)
Comment (DavidSubiros) [reply api/handlers.go:106]: Yes, absolutely right! - but if we did the whole loop we would update as many items as possible, and just ignore the ones that we can't be updated - I don't know what would be the best option, I will accept it as it is now.
